### ISSUE TITLE
Enable logit bias/grammar constraints for tool calling across all workflows

### ISSUE THREAD

**Issue by Soypete** [Soypete/PedroCLI#92]
## Problem

The codebase has a complete logit bias/grammar constraint system (`pkg/logits/`) but **it's not being used anywhere**. Instead, tools are permissive and use inference fallbacks, leading to:

- ⚠️ Warnings like `web_scraper: inferred action=scrape_url from args`
- Inconsistent tool call formats across models
- Missing required parameters only caught at execution time (not generation time)
- Tools need defensive coding with parameter inference logic

### ✅ Infrastructure Exists
- `pkg/logits/toolcall.go` - `ToolCallFilter` for enforcing `{"tool": "name", "args": {...}}` format
- `pkg/logits/grammar.go` - GBNF grammar generation from JSON schemas
- `pkg/llm/server.go:105-117` - Backend support for `logit_bias` and `grammar` parameters
- `pkg/llm/server.go:269-313` - **Tokenize() method already implemented** ✅
- `pkg/llm/interface.go` - `InferenceRequest.LogitBias` and `InferenceRequest.Grammar` fields

### ❌ Not Connected
- **Tokenize() not in Backend interface** - Exists in ServerClient but not exposed
- No agent sets `Grammar` or `LogitBias` in inference requests
- `LogitTool` exists but isn't registered with any agent
- All workflows rely on permissive tools with inference fallbacks

## Desired Behavior

**Use logit bias to enforce strict tool call format:**

```json
{
  "tool": "web_scraper",
  "args": {
    "action": "scrape_url",  // ← REQUIRED, enforced by logit bias
    "url": "https://example.com"
  }
}
```

Instead of current permissive behavior:
```json
{
  "tool": "web_scraper",
  "args": {
    "url": "https://example.com"  // ← Missing action, tool infers it
  }
}
```

### Phase 0: Wire Up Tokenize() (Prerequisite)

**File**: `pkg/llm/interface.go`

Add `Tokenize()` to Backend interface:

```go
type Backend interface {
    Infer(ctx context.Context, req *InferenceRequest) (*InferenceResponse, error)
    GetContextWindow() int
    GetUsableContext() int
    
    // Add tokenization support
    Tokenize(ctx context.Context, text string) ([]int, error)  // ← ADD THIS
}
```

**Why**: `ServerClient.Tokenize()` already exists (lines 269-313) but it's not exposed through the Backend interface, so agents can't use it.

**File**: `pkg/llm/ollama.go`

Add Tokenize() implementation for Ollama backend:

```go
func (o *OllamaClient) Tokenize(ctx context.Context, text string) ([]int, error) {
    // Ollama may not support /tokenize endpoint
    // Return error or implement via Ollama API
    return nil, fmt.Errorf("tokenization not supported by Ollama backend")
}
```

**File**: `pkg/llm/llamacpp.go`

Ensure LlamaCppClient implements Tokenize() (should inherit from ServerClient):

```go
// LlamaCppClient embeds ServerClient, so Tokenize() already available
```

### Phase 1: Blog Workflow (GLM-4.7)

**File**: `pkg/agents/blog_content.go`

1. **Add tokenizer to BlogContentAgent**:
   ```go
   type BlogContentAgent struct {
       backend   llm.Backend
       tokenizer logits.Tokenizer  // NEW
       // ...
   }
   ```

2. **Initialize tokenizer** (uses Backend.Tokenize()):
   ```go
   func NewBlogContentAgent(cfg AgentConfig) (*BlogContentAgent, error) {
       // Create tokenizer wrapper around backend
       tokenizer := &BackendTokenizer{backend: cfg.Backend}

       agent := &BlogContentAgent{
           backend:   cfg.Backend,
           tokenizer: tokenizer,
           // ...
       }
   }
   ```

3. **Create BackendTokenizer adapter**:
   ```go
   // pkg/logits/tokenizer.go
   type BackendTokenizer struct {
       backend llm.Backend
       vocab   map[int]string  // Cached vocabulary
   }

   func (t *BackendTokenizer) Tokenize(text string) ([]int, error) {
       return t.backend.Tokenize(context.Background(), text)
   }

   func (t *BackendTokenizer) GetVocabulary() map[int]string {
       // Build vocabulary by tokenizing known strings
       // Or fetch from backend if supported
       return t.vocab
   }

   func (t *BackendTokenizer) EOSToken() int {
       // Return EOS token ID (model-specific)
       return 2  // Common for many models
   }
   ```

4. **Enable ToolCallFilter in inference**:
   ```go
   // In executeInference() or similar
   toolDefs := convertToolsToLogitDefinitions(a.registry.ListTools())
   toolFilter, err := logits.NewToolCallFilter(toolDefs, a.tokenizer)
   if err != nil {
       return nil, fmt.Errorf("create tool filter: %w", err)
   }

   // Generate grammar string
   grammar, err := toolFilter.GenerateGrammar()
   if err != nil {
       return nil, fmt.Errorf("generate grammar: %w", err)
   }

   // Add to inference request
   req := &llm.InferenceRequest{
       SystemPrompt: systemPrompt,
       UserPrompt:   userPrompt,
       Temperature:  a.config.Model.Temperature,
       MaxTokens:    maxTokens,
       Grammar:      grammar,  // ← Enforce structure
   }
   ```

5. **Update web_scraper tool**:
   ```go
   // Remove inference logic from Execute()
   action, ok := args["action"].(string)
   if !ok || action == "" {
       // NO LONGER INFER - just fail
       return &Result{
           Success: false,
           Error:   "missing required parameter: action",
       }, nil
   }
   ```

### Phase 2: Grammar vs Tokenize Approach

**Two approaches available:**

#### Option A: Grammar Files (Recommended ⭐)

**Pros:**
- Simpler - no token ID management
- Server-side enforcement (llama-server has native GBNF support)
- No vocabulary caching needed
- Works with any backend that supports grammar

**Implementation:**
```go
grammar := generateToolCallGrammar(tools)  // Generate GBNF string
req.Grammar = grammar  // llama-server enforces at logit level
```

#### Option B: Tokenize + Logit Bias

**Pros:**
- Fine-grained control over token probabilities
- Can boost/suppress specific tokens
- Works with backends that don't support grammar

**Implementation:**
```go
// Tokenize allowed values
tokens, _ := backend.Tokenize(ctx, "\"web_scraper\"")

// Build logit bias map
req.LogitBias = map[int]float32{
    tokens[0]: 100.0,  // Boost allowed tokens
}
```

**Recommendation**: Use **Grammar approach** (Option A) as primary method, with **Tokenize + LogitBias** as fallback for backends without grammar support.

### Phase 3: Code Workflows (Builder, Debugger, Reviewer, Triager)

**File**: `pkg/agents/base.go`

Update `BaseAgent` to support logit bias:

```go
type BaseAgent struct {
    backend   llm.Backend
    registry  *tools.Registry
    tokenizer logits.Tokenizer  // NEW (optional, for logit bias fallback)
    // ...
}

func (a *BaseAgent) executeInference(...) (*llm.InferenceResponse, error) {
    // Try grammar approach first
    grammar, err := a.generateToolCallGrammar()
    if err == nil && grammar != "" {
        req.Grammar = grammar
    } else if a.tokenizer != nil {
        // Fallback to logit bias if grammar not available
        logitBias, err := a.generateLogitBias()
        if err == nil {
            req.LogitBias = logitBias
        }
    }

    return a.backend.Infer(ctx, req)
}
```

**Files to update**:
- `pkg/agents/builder.go`
- `pkg/agents/debugger.go`
- `pkg/agents/reviewer.go`
- `pkg/agents/triager.go`

All inherit from `BaseAgent`, so this should propagate automatically.

### Phase 4: Tool Schema Updates

**Define required parameters for all tools:**

**File**: `pkg/tools/webscraper.go`

```go
func (t *WebScraperTool) Schema() *logits.JSONSchema {
    return &logits.JSONSchema{
        Type: "object",
        Properties: map[string]*logits.JSONSchema{
            "action": {
                Type:        "string",
                Description: "Action to perform",
                Enum:        []string{"scrape_url", "scrape_local", "scrape_github"},
                // REQUIRED, enforced by grammar
            },
            "url": {
                Type:        "string",
                Description: "URL to scrape (required for scrape_url)",
            },
            "path": {
                Type:        "string",
                Description: "Local file path (required for scrape_local)",
            },
            "repo": {
                Type:        "string",
                Description: "GitHub repo (required for scrape_github)",
            },
        },
        Required: []string{"action"},  // ← Grammar enforces this
    }
}
```

**Similar updates needed for**:
- `pkg/tools/websearch.go`
- `pkg/tools/codeedit.go`
- `pkg/tools/file.go`
- `pkg/tools/bash.go`
- All other tools with optional-but-important parameters

### Phase 5: Testing

**Unit tests**:

```go
func TestBackend_Tokenize(t *testing.T) {
    client := NewServerClient("http://localhost:8082")
    
    tokens, err := client.Tokenize(context.Background(), "web_scraper")
    if err != nil {
        t.Fatalf("Tokenize failed: %v", err)
    }
    
    if len(tokens) == 0 {
        t.Error("Expected non-empty token list")
    }
}

func TestToolCallFilter_EnforcesAction(t *testing.T) {
    // Setup tokenizer and tool definitions
    backend := NewMockBackend()
    tokenizer := &BackendTokenizer{backend: backend}
    
    tools := []*logits.ToolDefinition{
        {
            Name: "web_scraper",
            Parameters: &logits.JSONSchema{
                Type: "object",
                Properties: map[string]*logits.JSONSchema{
                    "action": {Type: "string", Enum: []string{"scrape_url"}},
                    "url":    {Type: "string"},
                },
                Required: []string{"action"},
            },
        },
    }

    filter, err := logits.NewToolCallFilter(tools, tokenizer)
    if err != nil {
        t.Fatal(err)
    }

    // Test grammar generation
    grammar, err := filter.GenerateGrammar()
    if err != nil {
        t.Fatal(err)
    }
    
    if !strings.Contains(grammar, "action") {
        t.Error("Grammar should include 'action' field")
    }
}
```

**Integration test**:

```bash
# Run blog workflow with logit bias enabled
./pedrocli blog -file test.txt

# Verify:
# 1. No "inferred action" warnings
# 2. All tool calls have required parameters
# 3. Tool execution succeeds
```

### Phase 0: Wire Up Tokenize
- `pkg/llm/interface.go` - Add `Tokenize()` to Backend interface ⭐ **CRITICAL**
- `pkg/llm/ollama.go` - Implement Tokenize() (or return error if unsupported)

### Core Infrastructure
- `pkg/agents/base.go` - Add grammar/logit bias support to BaseAgent
- `pkg/agents/executor.go` - Enable grammar in InferenceExecutor
- `pkg/agents/blog_content.go` - Add tokenizer to BlogContentAgent
- `pkg/logits/tokenizer.go` - Create BackendTokenizer adapter

### Tool Updates (Remove Inference Logic)
- `pkg/tools/webscraper.go` - Remove action inference, add Schema()
- `pkg/tools/websearch.go` - Make parameters strict
- `pkg/tools/codeedit.go` - Enforce required fields
- `pkg/tools/file.go` - Strict parameter validation
- `pkg/tools/bash.go` - Strict parameter validation

### Tests
- `pkg/llm/server_test.go` - Test Tokenize() endpoint
- `pkg/agents/base_test.go` - Test grammar generation
- `pkg/agents/executor_test.go` - Test logit bias integration
- `pkg/tools/webscraper_test.go` - Test strict parameter validation

## Benefits

- ✅ **Eliminates warnings** - No more "inferred action" messages
- ✅ **Catches errors early** - Missing parameters detected at generation time, not execution
- ✅ **Consistent format** - All tool calls follow exact schema across all models
- ✅ **Simpler tools** - No defensive coding or inference logic needed
- ✅ **Better reliability** - Grammar constraints prevent malformed tool calls
- ✅ **Model agnostic** - Works with any backend that supports grammar constraints

## Implementation Sequence

1. **Phase 0** (5 min) - Add Tokenize() to Backend interface
2. **Phase 1** (2 hours) - Blog workflow + GLM-4.7
3. **Phase 2** (1 hour) - BackendTokenizer adapter
4. **Phase 4** (2 hours) - Tool schema definitions
5. **Phase 3** (1 hour) - Code workflows
6. **Phase 5** (1 hour) - Testing

**Total**: ~7 hours of work

### Tokenizer Vocabulary
- Need full vocabulary for logit-level control
- Options:
  1. **Use grammar** (recommended) - Server handles logit bias
  2. **Query /tokenize** for common strings, build vocab cache
  3. **Load from model metadata** if available

**Solution**: Use grammar approach (Option A) to avoid vocabulary management

### Backward Compatibility
- Older models without grammar support
- Fallback: Keep inference logic but add deprecation warnings

**Solution**: Make grammar optional, gracefully degrade:
```go
if grammar != "" {
    req.Grammar = grammar
} else {
    log.Warn("Grammar not available, tool calls may be inconsistent")
}
```

### Performance
- Grammar parsing adds overhead
- Tokenization required for logit bias fallback

**Mitigation**: Cache grammars, lazy-load tokenizer

## Success Criteria

- ✅ `Tokenize()` added to Backend interface and working
- ✅ Blog workflow uses grammar constraints with GLM-4.7
- ✅ All workflows (build, debug, review, triage) use grammar
- ✅ No "inferred action" warnings in logs
- ✅ Tools fail fast on missing required parameters
- ✅ Tool calls validated at generation time, not execution time
- ✅ Tests pass with strict parameter validation

## Related Issues

- #77 - BlogContentAgent improvements
- #91 - Multiple RSS feeds with podcast scraping
- `docs/issues/fix-web-scraper-action-parameter.md` - Action inference warning

## Priority

**Medium-High** - Improves reliability and consistency across all workflows, eliminates warnings, better error handling

## Labels

- `enhancement`
- `tool-calling`
- `logit-bias`
- `grammar`
- `glm-4`
- `blog`
- `workflows`
